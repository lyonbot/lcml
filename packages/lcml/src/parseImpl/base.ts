import { ParseError } from '../ParseError';
import type { StringStream } from '../StringStream';

const $$isParsedNode = Symbol();

export interface ParsedNodeBase {
  [$$isParsedNode]: true;
  type: string;
  start: number;
  end: number;
  parsedLength: number;
}

export function isParsedNode(x: any): x is ParsedNodeBase {
  return typeof x === 'object' && x !== null && x[$$isParsedNode] === true;
}

const defaultCtxOpts = {
  hook: null as ((node: ParsedNodeBase) => void) | null | undefined,
};
const ctxOpts = { ...defaultCtxOpts };

export function startParsing(opts: Partial<typeof ctxOpts>) {
  Object.assign(ctxOpts, defaultCtxOpts, opts);
  _isRecovering = null;
}
export function endParsing() {
  Object.assign(ctxOpts, defaultCtxOpts);
  _isRecovering = null;
}

// while recovering,
// array and object's parsing processes are stopped
// and partially parsed result is generated, returned

let _isRecovering: ParseError | null = null;
export const isRecovering = () => _isRecovering;

export function makeParsedNode<T extends ParsedNodeBase>(
  sourceStream: StringStream,
  extra: Omit<T, 'start' | 'end' | typeof $$isParsedNode>,
): T {
  const parsedLength = extra.parsedLength;

  const answer = {
    ...extra,
    start: sourceStream.pos,
    end: sourceStream.pos + parsedLength,
  } as T;

  Object.defineProperty(answer, $$isParsedNode, {
    configurable: true,
    enumerable: false,
    value: true,
  });

  if (ctxOpts.hook) ctxOpts.hook(answer);
  return answer;
}

/**
 * if current parsing process allows recovering, this set internal `isRecovering` flag to true, call `partialFinalizer()` and returns.
 *
 * otherwise, create a new ParseError and throw it
 */
export function makePanic<T extends ParsedNodeBase>(
  partialFinalizer: () => T,
  ...args: ConstructorParameters<typeof ParseError>
) {
  const error = new ParseError(...args);
  _isRecovering = error;
  return partialFinalizer();
  // throw error;
}
