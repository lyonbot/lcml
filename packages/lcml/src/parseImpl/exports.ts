export { ParseError } from '../ParseError';
export { isParsedNode, ParsedNodeBase } from './base';

import type { ParsedArrayNode } from './parseArray';
import type { ParsedCommentNode } from './parseComment';
import type { ParsedExpressionNode } from './parseExpression';
import type { ParsedBooleanNode, ParsedNullishNode, ParsedNumberNode, ParsedLiteralNode } from './parseLiteral';
import type { ParsedObjectPropertyNode, ParsedObjectNode } from './parseObject';
import type { ParsedStringNode, ParsedStringSegmentNode } from './parseString';
import type { ParsedIdentifierNode } from './parseIdentifier';
import type { ParsedValueNode } from './parseValue';

export type ParsedNode =
  | ParsedArrayNode
  | ParsedCommentNode
  | ParsedExpressionNode
  | ParsedBooleanNode
  | ParsedNullishNode
  | ParsedNumberNode
  | ParsedLiteralNode
  | ParsedObjectPropertyNode
  | ParsedObjectNode
  | ParsedStringNode
  | ParsedStringSegmentNode
  | ParsedIdentifierNode
  | ParsedValueNode;

export {
  ParsedArrayNode,
  ParsedCommentNode,
  ParsedExpressionNode,
  ParsedBooleanNode,
  ParsedNullishNode,
  ParsedNumberNode,
  ParsedLiteralNode,
  ParsedObjectPropertyNode,
  ParsedObjectNode,
  ParsedStringNode,
  ParsedStringSegmentNode,
  ParsedIdentifierNode,
  ParsedValueNode,
};
