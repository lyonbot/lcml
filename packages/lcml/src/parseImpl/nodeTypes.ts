import type { ParsedArrayNode } from './parseArray';
import type { ParsedCommentNode } from './parseComment';
import type { ParsedExpressionNode } from './parseExpression';
import type { ParsedBooleanNode, ParsedNullishNode, ParsedNumberNode, ParsedLiteralNode } from './parseLiteral';
import type { ParsedObjectPropertyNode, ParsedObjectNode } from './parseObject';
import type { ParsedStringNode, ParsedStringSegmentNode } from './parseString';
import type { ParsedIdentifierNode } from './parseIdentifier';
import type { ParsedValueNode } from './parseValue';

export interface ParsedNodeTypeLUT {
  'array': ParsedArrayNode;
  'comment': ParsedCommentNode;
  'expression': ParsedExpressionNode;
  'boolean': ParsedBooleanNode;
  'nullish': ParsedNullishNode;
  'number': ParsedNumberNode;
  // "literal": ParsedLiteralNode;
  'object-property': ParsedObjectPropertyNode;
  'object': ParsedObjectNode;
  'string': ParsedStringNode;
  'string-segment': ParsedStringSegmentNode;
  'identifier': ParsedIdentifierNode;
  // "value": ParsedValueNode;
}

type TempCheck1 = { [x in keyof ParsedNodeTypeLUT]: ParsedNodeTypeLUT[x] extends { type: x } ? never : x };
type BadItemsInLUT = TempCheck1[keyof TempCheck1];

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const _check__noBadItemsInLUT: BadItemsInLUT extends never ? true : false = true;

export type ParsedNode = ParsedNodeTypeLUT[keyof ParsedNodeTypeLUT];

export {
  ParsedArrayNode,
  ParsedCommentNode,
  ParsedExpressionNode,
  ParsedBooleanNode,
  ParsedNullishNode,
  ParsedNumberNode,
  ParsedLiteralNode,
  ParsedObjectPropertyNode,
  ParsedObjectNode,
  ParsedStringNode,
  ParsedStringSegmentNode,
  ParsedIdentifierNode,
  ParsedValueNode,
};
