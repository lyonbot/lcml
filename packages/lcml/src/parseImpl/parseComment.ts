import { ParsedNodeBase, makeParsedNode } from './base';
import { StringStream } from '../StringStream';

export interface ParsedCommentNode extends ParsedNodeBase {
  type: 'comment';
}

export function parseComment(stream: StringStream, top: string) {
  if (!(top === '/')) return;

  // maybe comment
  const lead = stream.peek(2);
  const end = lead === '//' ? ['\n', '\r'] : lead === '/*' ? ['*/'] : null;
  if (!end) return;

  // skip leading and find ending token
  // if ending token is not found, skip to the end of the stream
  const endToken = stream.indexOf(end, 2);

  return makeParsedNode<ParsedCommentNode>(stream, {
    type: 'comment',
    parsedLength: endToken ? endToken.endOffset : stream.remainderLength,
  });
}
