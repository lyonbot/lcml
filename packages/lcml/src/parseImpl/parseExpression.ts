import { ParsedNodeBase, makeParsedNode } from './base';
import { StringStream } from '../StringStream';

export interface ParsedExpressionNode extends ParsedNodeBase {
  type: 'expression';
  expression: string;
  expressionStart: number;
  expressionEnd: number;
}

export function parseExpression(stream: StringStream, top: string): ParsedExpressionNode | void {
  if (top !== '{') return;
  if (stream.peek(2) !== '{{') return;

  const endToken = stream.indexOf('}}', 2);
  if (!endToken) return;

  return makeParsedNode<ParsedExpressionNode>(stream, {
    type: 'expression',
    parsedLength: endToken.endOffset,
    expression: stream.substr(2, endToken.offset - 2),
    expressionStart: 2 + stream.pos,
    expressionEnd: endToken.offset + stream.pos,
  });
}
