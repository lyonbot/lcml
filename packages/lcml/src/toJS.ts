import { ParsedExpressionNode, ParsedNode, ParsedValueNode } from './parseImpl/exports';
import { unescapeString } from './string';

/**
 * toJS options
 *
 * @public
 */
export interface ToJSOptions {
  /**
   * In the generated code, you _might_ see `toString(something)` -- LCML implicitly calls a toString method.
   * This happens when user use `{{ expression }}` inside a string literal.
   *
   * If your `toString` has different name, provide it here.
   */
  globalToStringMethod?: string;
  /**
   * process expressions while generating js
   */
  processExpression?: null | ((node: ParsedExpressionNode, parents: ParsedNode[]) => string);
  /**
   * shortcut to set empty indent & lineBreak
   */
  compact?: boolean;
  /**
   * @default "  "
   */
  indent?: string;
  /**
   * this affects array and object
   *
   * @default "\n"
   */
  lineBreak?: string;
}

export function toJS(node: ParsedValueNode, opt: ToJSOptions = {}): string {
  const compact = !!opt.compact;
  const fullOpt: Required<ToJSOptions> = {
    globalToStringMethod: 'toString',
    processExpression: null,
    compact,
    indent: compact ? '' : '  ',
    lineBreak: compact ? ' ' : '\n',
    ...opt,
  };

  return internalToJS(node, '', fullOpt, () => []);
}

function internalToJS(
  node: ParsedValueNode,
  indent = '',
  opt: Required<ToJSOptions>,
  getParents: () => ParsedValueNode[],
): string {
  const nextParents = () => [...getParents(), node];
  const subCall = (node: ParsedValueNode) => internalToJS(node, opt.indent + indent, opt, nextParents);
  const type = node.type;

  switch (type) {
    case 'boolean':
    case 'number':
    case 'nullish':
      return node.raw;

    case 'expression':
      return `(${opt.processExpression ? opt.processExpression(node, getParents()) : node.expression})`;

    case 'string':
      return node.segments
        .map(x => {
          if (x.type === 'expression') return `${opt.globalToStringMethod}(${subCall(x)})`;
          if (!x.raw) return '';
          return JSON.stringify(unescapeString(x.raw));
        })
        .filter(Boolean)
        .join(' + ');

    case 'array':
      if (!node.length) return '[]';
      return ['[', ...node.items.map(x => opt.indent + (x ? subCall(x) : `/* empty */`) + ','), `]`].join(
        opt.lineBreak + indent,
      );

    case 'object':
      if (!node.properties.length) return '{}';
      return [
        '{',
        ...node.properties.map(x => {
          let key = '';

          if (x.key.type === 'identifier') key = x.key.raw;
          else if (x.key.type === 'expression' || x.key.isDynamic) key = '[' + subCall(x.key) + ']';
          else key = toJS(x.key);

          return `${opt.indent}${key}: ${x.value ? subCall(x.value) : 'undefined'},`;
        }),
        `}`,
      ].join(opt.lineBreak + indent);

    /* istanbul ignore next */
    default: {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const _: never = type;
      return '';
    }
  }
}
