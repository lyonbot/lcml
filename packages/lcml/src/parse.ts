import { parseComment } from './parseImpl/parseComment';
import { parseValue } from './parseImpl/parseValue';
import { ParseError } from './ParseError';
import { StringStream } from './StringStream';
import { endParsing, isRecovering, startParsing } from './parseImpl/base';
import { ParsedExpressionNode } from './parseImpl/exports';
import { internalParseStringContent } from './parseImpl/parseString';

/**
 * parser options
 *
 * @public
 */
export interface ParseOptions {
  /**
   * when input is empty, what shall be returned?
   *
   * @default "as-undefined"
   */
  treatEmptyInput?: 'as-undefined' | 'as-empty-string';

  /**
   * after parsing something, if there is still characters left unparsed, shall we just ignore them, or treat as error
   *
   * @default "error"
   */
  treatUnparsedRemainder?: 'as-error' | 'ignore';

  /**
   * when error occurs
   *
   * - `"throw"` (default) - a ParseError will be thrown
   * - `"recover"` - error and partial-parsed result will be return together without throwing
   * - `"as-string"` - discard all parsed structures and treat whole input as a string with expressions
   *
   * @default "throw"
   */
  onError?: 'throw' | 'recover' | 'as-string';
}

export type ParseResult = ReturnType<typeof parse>;

/**
 * skip leading comments and parse the value
 *
 * @public
 */
export function parse(str: string, opts: ParseOptions = {}) {
  const expressions = [] as ParsedExpressionNode[];

  try {
    startParsing({
      hook: node => {
        if (node.type === 'expression') expressions.push(node as ParsedExpressionNode);
      },
    });

    let stream = new StringStream(str);

    // skip leading comments

    while ((stream.skipSpaces(), !stream.eof())) {
      const top = stream.peek();

      const t = parseComment(stream, top);
      if (!t) break;
      stream.precede(t.parsedLength);
    }

    // start consuming value

    const actualStart = stream.pos;

    let error: ParseError | null = null;

    if (stream.eof()) {
      // see opts.emptyInput
      const placeholder = opts.treatEmptyInput === 'as-empty-string' ? '""' : 'undefined';
      stream = new StringStream(new Array(actualStart).fill(' ').join('') + placeholder);
      stream.precede(actualStart);
    }

    const top = stream.peek();
    let ast = parseValue(stream, top);

    if (!ast) {
      stream.precede(actualStart);
      ast = internalParseStringContent(stream, '')!; // empty input is processed above
    }

    // successfully parsed something
    stream.precede(ast.parsedLength);
    stream.skipSpaces();

    error = isRecovering();

    // do extra check here: is EOF reached?
    if (!error && !stream.eof()) {
      if (ast.type === 'expression') {
        // special case. see README.md
        const stream = new StringStream(str);
        stream.precede(actualStart);
        ast = internalParseStringContent(stream, '');
      } else if (opts.treatUnparsedRemainder !== 'ignore') {
        // by default, treat unparsed part as error
        error = new ParseError('unexpected remainder', stream);
      }
    }

    if (error && (opts.onError || 'throw') === 'throw') {
      throw error;
    }

    if (error && opts.onError === 'as-string') {
      // discard all parsed structures
      // treat whole input as string
      const stream = new StringStream(str);
      stream.precede(actualStart);
      ast = internalParseStringContent(stream, '');
    }

    return {
      ast,
      expressions,
      actualStart,
      end: actualStart + (ast ? ast.parsedLength : 0),
      error,
    };
  } finally {
    endParsing();
  }
}
