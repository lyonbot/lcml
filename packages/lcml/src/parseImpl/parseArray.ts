import { ParsedNodeBase, makeParsedNode, isRecovering, makePanic } from './base';
import { parseComment } from './parseComment';
import { StringStream } from '../StringStream';
import { parseValue, ParsedValueNode } from './parseValue';

export interface ParsedArrayNode extends ParsedNodeBase {
  type: 'array';
  length: number;
  items: (ParsedValueNode | null)[];
  itemsLocation: {
    hasValue: boolean;
    valueStart: number;
    valueEnd: number;

    hasTrailingComma?: boolean;
    trailingCommaStart?: number;
  }[];
}

export function parseArray(stream: StringStream, top: string): ParsedArrayNode | void {
  if (top !== '[') return;

  const ss = stream.clone(1);

  const items = [] as ParsedArrayNode['items'];
  const itemsLocation = [] as ParsedArrayNode['itemsLocation'];

  let index = 0;
  let slotEmpty = true;
  let finished = false;

  const finalize = () => {
    let length = itemsLocation.length;
    if (length > 2 && !itemsLocation[length - 1].hasValue) {
      // remove invalid empty slot made by trailing comma
      length--;
      itemsLocation.pop();
    }

    return makeParsedNode<ParsedArrayNode>(stream, {
      type: 'array',
      parsedLength: ss.pos - stream.pos,
      length,
      items: itemsLocation.map((_, i) => items[i] || null),
      itemsLocation,
    });
  };

  while (!isRecovering() && !finished && (ss.skipSpaces(), !ss.eof())) {
    const top = ss.peek();
    if (top === ']') {
      ss.precede(1);
      finished = true;
      break;
    }

    if (top === ',') {
      itemsLocation[index] = {
        ...(itemsLocation[index] || {
          hasValue: false,
          valueStart: ss.pos,
          valueEnd: ss.pos,
        }),
        hasTrailingComma: true,
        trailingCommaStart: ss.pos,
      };

      index++;
      slotEmpty = true;

      ss.precede(1);

      // after updating index,
      // prepare the next item's location info
      // this is useful for empty slot
      itemsLocation[index] = {
        hasValue: false,
        valueStart: ss.pos,
        valueEnd: ss.pos,
      };

      continue;
    }

    const comment = parseComment(ss, top);
    if (comment) {
      ss.precede(comment.parsedLength);
      continue;
    }

    const item = slotEmpty && parseValue(ss, top);
    if (!item) {
      return makePanic(finalize, `expect ${slotEmpty ? 'value, ' : ''}comma or right square bracket`, ss);
    }

    items[index] = item;
    itemsLocation[index] = {
      ...itemsLocation[index],
      hasValue: true,
      valueStart: ss.pos,
      valueEnd: ss.pos + item.parsedLength,
    };

    slotEmpty = false;
    ss.precede(item.parsedLength);
  }

  if (!isRecovering() && !finished) {
    return makePanic(finalize, 'expect right square bracket', ss);
  }

  return finalize();
}
