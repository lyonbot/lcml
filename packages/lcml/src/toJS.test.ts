import { parse } from './parse';
import { toJS, ToJSOptions } from './toJS';
import { ParsedExpressionNode } from './parseImpl/parseExpression';

describe('toJS', () => {
  const fn1 = jest.fn<string, [ParsedExpressionNode]>(node => node.expression);
  const cases: [ToJSOptions, (() => void)?][] = [
    [{ compact: true }],
    [{}],
    [{ globalToStringMethod: 'myToString' }],
    [
      { processExpression: fn1 },
      () => {
        expect(fn1).toBeCalledTimes(3);
        expect(fn1.mock.calls[0][0].expression).toEqual(' expr ');
        expect(fn1.mock.calls[1][0].expression).toEqual(' test ');
        expect(fn1.mock.calls[2][0].expression).toEqual(' sym ');
      },
    ],
  ];

  test.each(cases)('should work: %o', opt => {
    const source = `{ 
      "foo {{ expr }}": 1, 
      "bar": [{{ test }}, false,,],
      {{ sym }}: [2, 3, 4]
    }`;
    const ast = parse(source, { onError: 'recover' }).ast!;
    const js = toJS(ast, opt);

    expect(js).toMatchSnapshot();
  });
});
