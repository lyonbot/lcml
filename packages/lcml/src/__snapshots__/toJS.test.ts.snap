// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toJS should work: {
  processExpression: <ref *1> [Function: mockConstructor] {
    [length]: 1,
    [name]: 'mockConstructor',
    [prototype]: { [constructor]: [Circular *1] },
    _isMockFunction: true,
    getMockImplementation: [Function (anonymous)] { [length]: 0, [name]: '' },
    mock: [Getter/Setter],
    mockClear: [Function (anonymous)] { [length]: 0, [name]: '' },
    mockReset: [Function (anonymous)] { [length]: 0, [name]: '' },
    mockRestore: [Function (anonymous)] { [length]: 0, [name]: '' },
    mockReturnValueOnce: [Function (anonymous)] { [length]: 1, [name]: '' },
    mockResolvedValueOnce: [Function (anonymous)] { [length]: 1, [name]: '' },
    mockRejectedValueOnce: [Function (anonymous)] { [length]: 1, [name]: '' },
    mockReturnValue: [Function (anonymous)] { [length]: 1, [name]: '' },
    mockResolvedValue: [Function (anonymous)] { [length]: 1, [name]: '' },
    mockRejectedValue: [Function (anonymous)] { [length]: 1, [name]: '' },
    mockImplementationOnce: [Function (anonymous)] { [length]: 1, [name]: '' },
    mockImplementation: [Function (anonymous)] { [length]: 1, [name]: '' },
    mockReturnThis: [Function (anonymous)] { [length]: 0, [name]: '' },
    mockName: [Function (anonymous)] { [length]: 1, [name]: '' },
    getMockName: [Function (anonymous)] { [length]: 0, [name]: '' }
  }
} 1`] = `
"{
  [\\"foo \\" + toString(( expr ))]: 1,
  \\"bar\\": [  
  ( test ),  
  false,  
  /* empty */,  
],
  [( sym )]: [  
  2,  
  3,  
  4,  
],
}"
`;

exports[`toJS should work: { compact: true } 1`] = `"{ [\\"foo \\" + toString(( expr ))]: 1, \\"bar\\": [ ( test ), false, /* empty */, ], [( sym )]: [ 2, 3, 4, ], }"`;

exports[`toJS should work: { globalToStringMethod: 'myToString' } 1`] = `
"{
  [\\"foo \\" + myToString(( expr ))]: 1,
  \\"bar\\": [  
  ( test ),  
  false,  
  /* empty */,  
],
  [( sym )]: [  
  2,  
  3,  
  4,  
],
}"
`;

exports[`toJS should work: {} 1`] = `
"{
  [\\"foo \\" + toString( expr )]: 1,
  \\"bar\\": [  
  ( test ),  
  false,  
  /* empty */,  
],
  [( sym )]: [  
  2,  
  3,  
  4,  
],
}"
`;
